### PR TITLE
[Merged by Bors] - feat: port GroupTheory.FreeGroup

### PR DESCRIPTION
**Pull request by nomeata**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by nomeata** [comment]
Ok, the first 1100 lines are compiling. Now hitting
```
code generator does not support recursor 'List.recOn' yet, consider using 'match ... with' and/or structural recursion
```
but it’s gotten late, so I’ll stop now and go to bed. The PR is free for anyone to take over .

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 1]
```diff
@@ -0,0 +1,1433 @@
...
+
+This file defines free groups over a type. Furthermore, it is shown that the free group construction
+is an instance of a monad. For the result that `FreeGroup` is the left adjoint to the forgetful
+functor from groups to types, see `algebra/category/Group/adjunctions`.
+
+## Main definitions
+
+* `FreeGroup`/`FreeAddGroup`: the free group (resp. free additive group) associated to a type
+  `α` defined as the words over `a : α × bool` modulo the relation `a * x * x⁻¹ * b = a * b`.
+* `FreeGroup.mk`/`FreeAddGroup.mk`: the canonical quotient map `list (α × bool) → FreeGroup α`.
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:27]:
```suggestion
  `α` defined as the words over `a : α × Bool` modulo the relation `a * x * x⁻¹ * b = a * b`.
* `FreeGroup.mk`/`FreeAddGroup.mk`: the canonical quotient map `List (α × Bool) → FreeGroup α`.
```

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 2]
```diff
@@ -0,0 +1,1433 @@
...
+import Mathlib.Data.List.Sublists
+import Mathlib.Data.List.Basic
+import Mathlib.GroupTheory.Subgroup.Basic
+
+/-!
+# Free groups
+
+This file defines free groups over a type. Furthermore, it is shown that the free group construction
+is an instance of a monad. For the result that `FreeGroup` is the left adjoint to the forgetful
+functor from groups to types, see `algebra/category/Group/adjunctions`.
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:21]:
```suggestion
functor from groups to types, see `Algebra/Category/Group/Adjunctions`.
```

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 3]
```diff
@@ -0,0 +1,1433 @@
...
+* `FreeGroup`/`FreeAddGroup`: the free group (resp. free additive group) associated to a type
+  `α` defined as the words over `a : α × bool` modulo the relation `a * x * x⁻¹ * b = a * b`.
+* `FreeGroup.mk`/`FreeAddGroup.mk`: the canonical quotient map `list (α × bool) → FreeGroup α`.
+* `FreeGroup.of`/`FreeAddGroup.of`: the canonical injection `α → FreeGroup α`.
+* `FreeGroup.lift f`/`FreeAddGroup.lift`: the canonical group homomorphism `FreeGroup α →* G`
+  given a group `G` and a function `f : α → G`.
+
+## Main statements
+
+* `free_group.church_rosser`/`free_add_group.church_rosser`: The Church-Rosser theorem for word
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:34]:
```suggestion
* `FreeGroup.Red.church_rosser`/`FreeAddGroup.Red.church_rosser`: The Church-Rosser theorem for word
```

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 4]
```diff
@@ -0,0 +1,1433 @@
...
+
+@[to_additive]
+theorem map_eq_lift : map f x = lift (of ∘ f) x :=
+  Eq.symm <| map.unique _ fun x => by simp
+#align free_group.map_eq_lift FreeGroup.map_eq_lift
+#align free_add_group.map_eq_lift FreeAddGroup.map_eq_lift
+
+/-- Equivalent types give rise to multiplicatively equivalent free groups.
+
+The converse can be found in `group_theory.free_abelian_group_finsupp`,
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:854]:
```suggestion
The converse can be found in `GroupTheory.FreeAbelianGroupFinsupp`,
```

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 5]
```diff
@@ -0,0 +1,1433 @@
...
+as `equiv.of_free_group_equiv`
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:855]:
```suggestion
as `Equiv.of_freeGroupEquiv`
```

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 6]
```diff
@@ -991,82 +1009,83 @@ variable {β : Type u}
...
     (Cm : ∀ x y, C x → C y → C (x * y)) : C z :=
   Quot.inductionOn z fun L =>
-    List.recOn L C1 fun ⟨x, b⟩ tl ih => Bool.recOn b (Cm _ _ (Ci _ <| Cp x) ih) (Cm _ _ (Cp x) ih)
+    List.recOn L C1 fun ⟨x, b⟩ _tl ih => Bool.recOn b (Cm _ _ (Ci _ <| Cp x) ih) (Cm _ _ (Cp x) ih)
 #align free_group.induction_on FreeGroup.induction_on
-#align free_add_group.induction_on freeAddGroup.induction_on
+#align free_add_group.induction_on FreeAddGroup.induction_on
 
-@[simp, to_additive]
+@[to_additive]
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:1025]: If you remove a simp attribute, please leave a porting note explaining why (a 1 line summary of the linter complaint).

File: Mathlib/GroupTheory/FreeGroup.lean [hunk 7]
```diff
@@ -991,82 +1009,83 @@ variable {β : Type u}
...
-@[simp, to_additive]
+@[to_additive (attr:=simp)]
 theorem map_inv (f : α → β) (x : FreeGroup α) : f <$> x⁻¹ = (f <$> x)⁻¹ :=
   (map f).map_inv x
 #align free_group.map_inv FreeGroup.map_inv
-#align free_add_group.map_neg freeAddGroup.map_neg
+#align free_add_group.map_neg FreeAddGroup.map_neg
 
-@[simp, to_additive]
+@[to_additive]
```
Comment (jcommelin) [inline Mathlib/GroupTheory/FreeGroup.lean:1049]: Idem

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**